### PR TITLE
[webview_flutter] Add listener for scroll position change in Android

### PR DESCRIPTION
**Pull request by TheVinhLuong**
## Description
Adding scroll listener and ability to scroll programmatically in Android.

## Related Issues
https://github.com/flutter/flutter/issues/42442
https://github.com/flutter/flutter/issues/31027

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by amirh** [comment]
Thanks for the contribution!

I'm following the initial PR review policy, this PR isn't trivial to review so I'm labeling it with "backlog" and we will prioritize according to the issue's priority.

Relevant issue: https://github.com/flutter/flutter/issues/42442

**Comment by andzejsw** [comment]
@amirh any progress on this?

**Comment by andzejsw** [comment]
@amirh any progress?

**Comment by TheVinhLuong** [comment]
@andzejsw In the meantime, you can try out this [plugin](https://pub.dev/packages/flutter_webview_plugin), it has an implemented scroll listener

**Review by real1900 — APPROVED** [review]

**Comment by stuartmorgan** [comment]
I took an initial look at this; at a high level it looks good, but per this is a case where we should land it in parallel with an iOS implementation per the policy described at https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#platform-support

If anyone is interested in moving this forward, I encourage you to open a PR based on this one that adds an iOS implementation; once we have both we can review this one and the iOS portion of that one.

This will also need to be updated for the federation changes (which is likely currently blocked on https://github.com/flutter/flutter/issues/94051 for the moment; we hope to resolve that at the beginning of 2022).

**Comment by stuartmorgan** [comment]
@TheVinhLuong Are you planning on updating this for the structural changes in the plugin (federation, Android conversion to a more Dart-based implementation), so that it would be ready as the foundation for a cross-platform PR?

**Comment by TheVinhLuong** [comment]
@stuartmorgan Sorry for my late reply. I will update the PR for the Android side, also for iOS if I can make it.

**Comment by Hixie** [comment]
@TheVinhLuong Just a friendly check-in to see if this is still something you have on your radar?

**Comment by TheVinhLuong** [comment]
> @TheVinhLuong Just a friendly check-in to see if this is still something you have on your radar?

Yes, I have implemented it both for android and ios. I am writing test now.

**Comment by TheVinhLuong** [comment]
@Hixie Could you please review my PR?

**Comment by bparrishMines** [comment]
@TheVinhLuong @Hixie We are currently working on a new interface for `webview_flutter` and aren't accepting new features until it is released.

See https://github.com/flutter/flutter/issues/94051
https://github.com/flutter/flutter/issues/93732

**Comment by stuartmorgan** [comment]
https://github.com/flutter/flutter/issues/93732 is resolved, but this is still blocked on us finishing https://github.com/flutter/flutter/issues/94051

**Comment by stuartmorgan** [comment]
Status update from triage: this is still blocked, but we hope to have it unblocked soon.

**Comment by stuartmorgan** [comment]
webview_flutter v4 has finally been completed, which included fixing the platform interface to that we can make additions to it without breaking changes. That means that this is finally unblocked! Thanks for your patience while we worked through the restructuring.

webview_flutter v4 uses a Dart-first implementation, where the platform channel boundary is as close as possible to the underlying host APIs, and plugin logic is written primarily in Dart rather than the host language. That means that this PR will need to be updated to follow that model. Please let us know if you have questions about updating to v4 (and https://github.com/flutter/plugins/pull/6881 may be useful as an example of adding a feature in the v4 structure).

**Comment by TheVinhLuong** [comment]
Thanks for the update. I will keep you updated on the progress of this PR.

**Comment by TheVinhLuong** [comment]
Hi @stuartmorgan, can you review my PR? I have updated the PR according to the new plugin federation format. 

Since this PR is just an implementation for the Android side the automated test for changelogs is failing for other platforms. Please add an 'override: no versioning needed' label for this PR if there is nothing wrong. Thank you!

**Comment by stuartmorgan** [comment]
> Since this PR is just an implementation for the Android side the automated test for changelogs is failing for other platforms. Please add an 'override: no versioning needed' label for this PR if there is nothing wrong.

You just need to revert the tool-generated pubspec changes in those other packages.

I'll review when I get a chance, but in the meantime you have Dart unit test and analysis failures in CI that will need to be addressed.

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java:9]: Instead of adding a new method to `WebView`, you should add a n`onScrollChanged` callback to the Dart `WebView`. The same way `WebViewClient` adds a callback: https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart#L759

I would also make the callback include the `WebView` for convenience:

```dart
final void Function(WebView webView, int left, int top, int oldLeft, int oldTop)? onScrollChanged;
```

File: packages/webview_flutter/webview_flutter_android/CHANGELOG.md
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/CHANGELOG.md:3]: Ios can also support listening to scroll events, so we should add the method to the platform interface and implement for both platforms. This will probably change to something like "Implements support for `PlatformWebViewController.setOnScrollChange`".

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:264]: To follow the current convention of the interface, this should probably be `setOnScrollChanged(...)` or `setOnContentOffsetChanged(...)`. I'm impartial to either of these.

**Review by bparrishMines — COMMENTED** [review]
Left some early feedback. Are you able to do the iOS portion as well or would you like help with that?

**Comment by ditman** [comment]
Sadly, I don't think this feature is currently implementable in the web, since there's no communication mechanism between the iframe that is being injected and the flutter web app.

**Comment by TheVinhLuong** [comment]
> Left some early feedback. Are you able to do the iOS portion as well or would you like help with that?

I think I can do it. Can I open a new PR for the iOS implementation?

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java
Comment (TheVinhLuong) [reply packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java:9]: Do you mean `WebViewExtendedApi` is not necessarily needed? I did read how the `WebViewClient` and attempted to do the same but couldn't find another way to set a scroll listener to a `WebView` without defining the `WebViewExtendedApi `.

The difference between setting a `ScrollListener` and a `WebViewClient` is that the `webkit.WebView` expose the method `setWebViewClient` for public use regardless of the Android required API level, this is unlike `setOnScrollChangeListener` since this is the method required min API level 23. The workaround here is to override the method`onScrollChanged` of the `WebView` and get the offset changed event from there.

Do you have another approach for this?

Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java:9]: Yea you can start by adding the `onScrollChanged` parameter to `WebView`:

```dart
class WebView {
  WebView({this.useHybridComposition = false, this.onScrollChanged}) : super.detached() {
    api.createFromInstance(this);
  }

  final void Function(WebView webView, int left, int top, int oldLeft, int oldTop)? onScrollChanged;
}
```

Then will need to establish a Flutter API for WebView in the `pigeons/android_webview.dart` file:

```dart
@FlutterApi()
abstract class WebViewFlutterApi {
  void onScrollChanged(int identifier, ...);
}
```

Then you can run the generator and implement the Flutter API that is generated. The implementation goes in `lib/src/android_webview_api_impl.dart`.

Then in Java you'll need to override the `onScrollChanged` method in the `WebViewPlatformView` class found in `WebViewHostApiImpl.java` class. You can then instantiate a `WebViewFlutterApi` and call the `WebViewFlutterApi.onScrollChanged` method in this overridden method.  

I may have missed a few things in this explanation, so definitely feel free to ask me for more clarification. The plugin is still suffering from tech debt, so adding new features still requires more work than they will in the future. I can also contribute code to this PR if needed.

**Comment by bparrishMines** [comment]
This PR should follow https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins and include the iOS implementation in this PR. Then once this has been approval, you can split the code from each package into their own PRs.

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java
Comment (TheVinhLuong) [reply packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewExtendedApi.java:9]: Thank you for the feedback. I did as you said but even though it helped to make things simple I still end up needing to retain `WebViewExtendedApi `.

**Comment by TheVinhLuong** [comment]
@bparrishMines Thanks for your feedback. Could you please take a look again at Android implementations for this? If no problems arise I will apply the same logic for the iOS side.

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!

**Pull request closed without merging**
